irc.go

```diff
@@ -176,11 +176,58 @@ var stdMemberships = []membership{
 	{'v', '+'}, // voice
 }
 
-func (m *membership) String() string {
-	if m == nil {
-		return ""
+// memberships always sorted by descending membership rank
+type memberships []membership
+
+func (m memberships) Add(availableMemberships []membership, newMembership membership) memberships {
+	i := 0
+	for _, availableMembership := range availableMemberships {
+		if i >= len(m) {
+			break
+		}
+		if m[i] == availableMembership {
+			if availableMembership == newMembership {
+				// we already have this membership
+				return m
+			}
+			i++
+			continue
+		}
+		if availableMembership == newMembership {
+			break
+		}
+	}
+	newMemberships := make([]membership, len(m)+1)
+	copy(newMemberships[:i], m[:i])
+	newMemberships[i] = newMembership
+	copy(newMemberships[i+1:], m[i:])
+	return newMemberships
+}
+
+func (m memberships) Remove(oldMembership membership) memberships {
+	for i, currentMembership := range m {
+		if currentMembership == oldMembership {
+			newMemberships := make([]membership, len(m)-1)
+			copy(newMemberships[:i], m[:i])
+			copy(newMemberships[i:], m[i+1:])
+			return newMemberships
+		}
+	}
+	return m
+}
+
+func (m memberships) String(dc *downstreamConn) string {
```

Comment:
Now that this doesn't implement `fmt.Stringer` anymore, maybe we should rename this to `MarshalPrefixes`

Reply:
what about `Format()`, like we do for channel modes?